runtime/lib/backend/LightningKokkosSimulator.cpp

```diff
@@ -245,81 +256,99 @@ auto LightningKokkosSimulator::Var(ObsIdType obsKey) -> double
     return m.var(*obs);
 }
 
-auto LightningKokkosSimulator::State() -> std::vector<std::complex<double>>
+void LightningKokkosSimulator::State(MemRefView<std::complex<double>, 1> &state)
 {
+    RT_ASSERT(!state.empty());
     const size_t num_qubits = this->device_sv->getNumQubits();
     const size_t size = Pennylane::Lightning_Kokkos::Util::exp2(num_qubits);
+    RT_FAIL_IF(state.size() != size, "Invalid size for the pre-allocated state vector");
 
-    std::vector<std::complex<double>> state(size);
-    auto *state_kptr = reinterpret_cast<Kokkos::complex<double> *>(state.data());
+    auto *state_kptr = reinterpret_cast<Kokkos::complex<double> *>(state.get().data_allocated);
```

Comment:
Doesn't this completely break the abstraction we are trying establish? 🤔  It directly accesses the raw pointer to the data buffer stored in the MemRef.

Reply:
Maybe there is no way around it though when dealing with device memory (copying into a host buffer and then populating the memref that is).

Reply:
Yes. I agree. It would be nice to have this support where if state is guaranteed to be contiguous and restricted (no other pointer to it exists), then we could get rid of the `std::vector<std::complex<double>> state(size)`.

I do believe that this is a guarantee we hold at the moment. Every time we lower the `quantum.state` to LLVM-IR we go through a process where we allocate an empty tensor which we pass to quantum state function call `__quantum__qis__state`. (Although I would change it so that instead of pointing to `state.get().data_allocated` I would point it to `(char*)state.get().data_aligned + state.get().offset`.

@dime10 I believe your concerns come from the one shot bufferizer. I do think we would be able to specify how we want `quantum.state` to be bufferized even with the one shot bufferizer. If we specify the bufferization stage to always create a new memref, then this should be safe.

Reply:
Agreed. This implementation would avoid the unnecessary copy of the memref buffer, but unfortunately doesn't yet take `strides` into account as there are no ways of using the iterator in the Kokkos::deep_copy method :/

To fix this, My proposal is by relying on `std::vector<std::comlex<double>> tmp_state(size)` and update the`state` by `std::move`.

Reply:
> I think however that this assumptions are not something that the writers of devices should know though, and they perhaps should stick to iterating over the memref view and assigning values to it?

This is my primary concern, devices shouldn't even care about what a memref is, much less make assumptions about the memory layout that the compiler may or may not guarantee 🤔 

> To fix this, I'm going to use std::vector<std::comlex<double>> buffer(size) and update the state by std::move.

I think this is a reasonable solution for devices with external memory in order to allow them to transfer their data as efficiently as possible. We can revisit the idea of the runtime providing a guarantee that the destination buffer conforms to certain expectations if we find that this is needed or beneficial enough.